### PR TITLE
 [TECHDEBT][Shared] Address legacy Run golangci-lint errors (Issue #295)

### LINKED ISSUE #295: [TECHDEBT][Shared] Address legacy Run golangci-lint errors
## Objective

Improve unit tests by addressing legacy linting errors 

## Origin Document

Raised by review of https://github.com/pokt-network/pocket/pull/235#issuecomment-1270639788

Screen_Shot_2022-10-06_at_10 50 28_AM

## Goals

- Block PRs that break unit tests
- Block PRs that break linting errors
- Display code changes that do not follow style guidelines 

## Deliverables

- [ ] A PR that tends to all of the legacy linting errors in the repo
- [ ] A PR that updates the linting github workflow to block (rather than simply display a warning) on golang and repo linting errors

## Non-goals / Non-deliverables

- Changing any business logic in the codebase

## General issue deliverables

- [ ] Update the appropriate CHANGELOG
- [ ] Update any relevant READMEs (local and/or global)
- [ ] Update any relevant global documentation & references
- [ ] If applicable, update the source code tree explanation
- [ ] If applicable, add or update a state, sequence or flowchart diagram using [mermaid](https://mermaid-js.github.io/mermaid/)

---

**Creator**: @jessicadaugherty
**Co-Owners**: @Olshansk

### PR DESCRIPTION
**Pull request by h5law**
## Description

This PR addresses all the errors from running `golangci-lint run`. The errors which were genuine have been addressed and those that are `false positives` have been ignored with the relevant `//nolint:{linter}` comments with their required justifications.

The `golangci-lint` github worklow has also been separated into a separate job and file with its own setup process which should counteract the `Cannot open file: File exists` errors

## Issue

Fixes #295 
## Type of change

Please mark the relevant option(s):

- [ ] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other 

## List of changes

- Fix legacy issues from linter
- Add comments to ignore false positives
- Seperate `golanci-lint` into its own worklow

## Testing

- [x] `make develop_test`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [x] I have updated the corresponding CHANGELOG

### If Applicable Checklist
- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

File: .github/workflows/golangci-lint.yaml [hunk 1]
```diff
@@ -0,0 +1,52 @@
+on:
```
Comment (Olshansk) [inline .github/workflows/golangci-lint.yaml:1]: Why did you move this out into a separate workflow? My only concern is the extra time related to installing dependencies and such.

File: app/client/cli/debug.go [hunk 2]
```diff
@@ -81,7 +81,10 @@ func NewDebugCommand() *cobra.Command {
 			}
 			p2pMod = p2pM.(modules.P2PModule)
 
-			p2pMod.Start()
+			err = p2pMod.Start()
+			if err != nil {
+				log.Fatalf("[ERROR] Failed to start p2p module: %v", err.Error())
+			}
```
Comment (Olshansk) [inline app/client/cli/debug.go:87]: Single line

File: app/client/cli/utils.go [hunk 3]
```diff
@@ -147,8 +152,15 @@ func postRawTx(ctx context.Context, pk crypto.Ed25519PrivateKey, j []byte) (*rpc
 }
 
 func getNonce() string {
-	rand.Seed(time.Now().UTC().UnixNano())
-	return fmt.Sprintf("%d", rand.Uint64())
+	max := new(big.Int)
```
Comment (Olshansk) [inline app/client/cli/utils.go:155]: There's probably an opportunity here to move random number generation (with fallback) into `shared`. Potentially `shared/crypto/rand.go`?

File: consensus/debugging.go [hunk 4]
```diff
@@ -42,18 +44,24 @@ func (m *consensusModule) GetNodeState() typesCons.ConsensusNodeState {
...
 	m.ResetForNewHeight()
 	m.clearLeader()
 	m.clearMessagesPool()
-	m.GetBus().GetPersistenceModule().HandleDebugMessage(&messaging.DebugMessage{
+	err := m.GetBus().GetPersistenceModule().HandleDebugMessage(&messaging.DebugMessage{
 		Action:  messaging.DebugMessageAction_DEBUG_PERSISTENCE_RESET_TO_GENESIS,
 		Message: nil,
 	})
-	m.GetBus().GetPersistenceModule().Start() // reload genesis state
+	if err != nil {
```
Comment (Olshansk) [inline consensus/debugging.go:58]: single line

File: utility/types/vote_test.go [hunk 5]
```diff
@@ -11,36 +11,38 @@ func TestVoteValidateBasic(t *testing.T) {
...
-	v3.BlockHash = []byte("not_a_hash")
+	v3 := createLegacyVote(publicKey.Bytes(), []byte("not_a_hash"), 1, DoubleSignEvidenceType)
 	badBlockHashLen := len(v3.BlockHash)
 	require.Equal(t, v3.ValidateBasic(), ErrInvalidHashLength(crypto.ErrInvalidHashLen(badBlockHashLen)))
 	// no hash
 	v3.BlockHash = nil
 	require.Equal(t, v3.ValidateBasic(), ErrEmptyHash())
 	// negative height
-	v4 := v
+	v4 := createLegacyVote(publicKey.Bytes(), testHash, -1, DoubleSignEvidenceType)
```
Comment (Olshansk) [inline utility/types/vote_test.go:31]: why not do just `proto.Clone(v)`?

File: utility/session.go [hunk 6]
```diff
@@ -106,12 +106,14 @@ func (s *session) findClosestYFishermen() []*coreTypes.Actor {
 // Q) why do we hash to find a newKey between every actor selection?
 // A) pseudo-random selection only works if each iteration is re-randomized
 //    or it would be subject to lexicographical proximity bias attacks
-func (s *session) pseudoRandomSelection(orderedListOfPublicKeys []string, numberOfActorsInSession int) []*coreTypes.Actor {
-	// IMPORTANT:
-	// THIS IS A DEMONSTRABLE FUNCTION THAT WILL NOT BE IMPLEMENTED AS SUCH
-	// IT EXISTS IN THIS COMMIT PURELY TO COMMUNICATE THE EXPECTED BEHAVIOR
-	return nil
-}
+/*
```
Comment (Olshansk) [inline utility/session.go:109]: Was this error linting on unused?

File: runtime/test_artifacts/util.go [hunk 7]
```diff
@@ -65,7 +65,9 @@ func SetupPostgresDocker() (*dockertest.Pool, *dockertest.Resource, string) {
 		}
 	}()
 
-	resource.Expire(1200) // Tell docker to hard kill the container in 20 minutes
+	if err := resource.Expire(1200); err != nil { // Tell docker to hard kill the container in 20 minutes
```
Comment (Olshansk) [inline runtime/test_artifacts/util.go:68]: Can you move 1200 into a constant at the top of the file?

File: persistence/validator.go [hunk 8]
```diff
@@ -8,11 +8,12 @@ import (
 	"github.com/pokt-network/pocket/shared/modules"
 )
 
-func (p PostgresContext) GetValidatorExists(address []byte, height int64) (exists bool, err error) {
+func (p *PostgresContext) GetValidatorExists(address []byte, height int64) (exists bool, err error) {
 	return p.GetExists(types.ValidatorActor, address, height)
 }
 
-func (p PostgresContext) GetValidator(address []byte, height int64) (operator, publicKey, stakedTokens, serviceURL, outputAddress string, pausedHeight, unstakingHeight int64, err error) {
+//nolint:gocritic
```
Comment (Olshansk) [inline persistence/validator.go:15]: Can you add a comment on the lint error here?

File: persistence/validator.go [hunk 9]
```diff
@@ -46,76 +47,53 @@ func (p PostgresContext) UpdateValidator(address []byte, serviceURL string, stak
...
 }
 
-func (p PostgresContext) SetValidatorPauseHeight(address []byte, height int64) error {
+func (p *PostgresContext) SetValidatorPauseHeight(address []byte, height int64) error {
 	return p.SetActorPauseHeight(types.ValidatorActor, address, height)
 }
 
-// TODO(team): The Get & Update operations need to be made atomic
-// TODO(team): Deprecate this functiona altogether and use UpdateValidator where applicable
-func (p PostgresContext) setValidatorStakedTokens(address []byte, tokens string) error {
```
Comment (Olshansk) [inline persistence/validator.go:83]: Was this function unused?

File: persistence/test/setup_test.go [hunk 10]
```diff
@@ -331,7 +340,7 @@ func getRandomServiceURL() string {
 }
 
 func getRandomBigIntString() string {
-	return converters.BigIntToString(big.NewInt(rand.Int63()))
+	return converters.BigIntToString(big.NewInt(rand.Int63())) //nolint:gosec // G404 - Weak random source is okay here
```
Comment (Olshansk) [inline persistence/test/setup_test.go:343]:
```suggestion
	return converters.BigIntToString(big.NewInt(rand.Int63())) //nolint:gosec // G404 - Weak random source is okay in unit tests
```

File: persistence/test/setup_test.go [hunk 11]
```diff
@@ -115,17 +118,21 @@ func newTestPersistenceModule(databaseUrl string) modules.PersistenceModule {
...
-		log.Fatalf("Error creating persistence module: %s", err)
+		log.Printf("Error creating persistence module: %s", err)
+		return nil
 	}
+
 	return persistenceMod.(modules.PersistenceModule)
 }
 
 // IMPROVE(team): Extend this to more complex and variable test cases challenging & randomizing the state of persistence.
+//nolint:gosec // G404 - Weak random source is okay here
```
Comment (Olshansk) [inline persistence/test/setup_test.go:135]:
```suggestion
//nolint:gosec // G404 - Weak random source is okay in unit tests
```

File: persistence/fisherman.go [hunk 12]
```diff
@@ -8,11 +8,12 @@ import (
 	"github.com/pokt-network/pocket/shared/modules"
 )
 
-func (p PostgresContext) GetFishermanExists(address []byte, height int64) (exists bool, err error) {
+func (p *PostgresContext) GetFishermanExists(address []byte, height int64) (exists bool, err error) {
 	return p.GetExists(types.FishermanActor, address, height)
 }
 
-func (p PostgresContext) GetFisherman(address []byte, height int64) (operator, publicKey, stakedTokens, serviceURL, outputAddress string, pausedHeight, unstakingHeight int64, chains []string, err error) {
+//nolint:gocritic
```
Comment (Olshansk) [inline persistence/fisherman.go:15]: Can you add comments w/ explanation to empty nolint

File: persistence/state.go [hunk 13]
```diff
@@ -140,13 +140,7 @@ func (p *PostgresContext) updateMerkleTrees() (string, error) {
...
 		switch treeType {
 		// Actor Merkle Trees
-		case appMerkleTree:
-			fallthrough
-		case valMerkleTree:
-			fallthrough
-		case fishMerkleTree:
-			fallthrough
-		case serviceNodeMerkleTree:
+		case appMerkleTree, valMerkleTree, fishMerkleTree, serviceNodeMerkleTree:
```
Comment (Olshansk) [inline persistence/state.go:143]: TIL!

File: p2p/raintree/network.go [hunk 14]
```diff
@@ -246,21 +250,19 @@ func (n *rainTreeNetwork) GetBus() modules.Bus {
 	return n.bus
 }
 
+// Generate cryptographically secure random nonce
 func getNonce() uint64 {
```
Comment (Olshansk) [inline p2p/raintree/network.go:254]: I left another comment about this elsewhere, but let's move out the helper into a shared utility under the crypto package.

File: p2p/providers/addrbook_provider/persistence/provider.go [hunk 15]
```diff
@@ -12,8 +12,8 @@ var _ modules.IntegratableModule = &persistenceAddrBookProvider{}
 var _ addrbook_provider.AddrBookProvider = &persistenceAddrBookProvider{}
 
 type persistenceAddrBookProvider struct {
-	bus         modules.Bus
-	p2pCfg      *configs.P2PConfig
+	bus modules.Bus
+	// p2pCfg      *configs.P2PConfig
```
Comment (Olshansk) [inline p2p/providers/addrbook_provider/persistence/provider.go:16]: Remove if unused

**Review by Olshansk — CHANGES_REQUESTED** [review]
Just a few minor comments.

Thanks for doing this. It's going to go a long way in catching bugs we won't have to encounter :)

File: .github/workflows/golangci-lint.yaml (hunk 1 above)
Comment (h5law) [reply .github/workflows/golangci-lint.yaml:1]: @Olshansk Reading through the link @okdas posted in issue #295 I saw this https://github.com/golangci/golangci-lint-action/issues/23#issuecomment-647177770 which says to separate the workflow into its own job - i went one further into its own file - I can put it back in but it needs its own job so that the errors like the ones below don't happen anymore

```bash
Cannot open: File exists
```

File: persistence/validator.go (hunk 9 above)
Comment (h5law) [reply persistence/validator.go:83]: @Olshansk yeah this was unused. Should I add it back?

File: utility/session.go (hunk 6 above)
Comment (h5law) [reply utility/session.go:109]: @Olshansk yeah this was unused - before I found out about `//nolint` so I can add it back in?

File: utility/types/vote_test.go (hunk 5 above)
Comment (h5law) [reply utility/types/vote_test.go:31]: With:

```go
v2 := proto.Clone(v)
v2.PublicKey = []byte("not_a_public_key")
```

I would get this error 

image

I am sure there is a better way around this error but I couldn't find much help online so I went with new `*LegacyVote` structs for each test

**Comment by h5law** [comment]
@Olshansk I have merged with main and addressed all linter errors from this merge. I also checked back and what I thought was gained from moving the workflow into its own file was not actually the case - the `Cannot open: File exists` errors I mean. So I have brought it back into the same test workflow job.

File: utility/types/vote_test.go (hunk 5 above)
Comment (Olshansk) [reply utility/types/vote_test.go]: I didn't test this out myself, but I think the reasoning for this might be because `proto.Clone` returns a generic `proto.Message`, but `PublicKey` is a property of the `LegacyVote` proto.

What if you type cast it first: `v2 := proto.Clone(v).(LegacyVote)`?

File: utility/session.go (hunk 6 above)
Comment (Olshansk) [reply utility/session.go]: Yea, let's add it back with a `not used but included for future implementation reference`

File: persistence/validator.go (hunk 9 above)
Comment (Olshansk) [reply persistence/validator.go:83]: Nah, if it's locally scoped, usused, and not needed for documentation purposes, it's increasing cognitive load.

We'll add it back if we ever need it.

File: utility/types/vote_test.go (hunk 5 above)
Comment (h5law) [reply utility/types/vote_test.go]: @Olshansk 
image

I tried this I think its because `proto.Clone()` doesn't return a generic message but a reflection of one which must behave differently I tried reading the docs about it but didn't make much progress.

File: .github/workflows/golangci-lint.yaml (hunk 1 above)
Comment (Olshansk) [reply .github/workflows/golangci-lint.yaml]: Makes sense. Can you just add a comment above the `job` field explaining why this is a separate job in its own file? You can probably repurpose most of your comment above.

Comment (Olshansk) [reply .github/workflows/golangci-lint.yaml]: Saw your comment. Ignore this & resolve.

Comment (h5law) [reply .github/workflows/golangci-lint.yaml]: @Olshansk I left a new comment below this was the intention however I double checked and this didn't actually solve those errors they still occurred. I have since moved the job back into the test job only difference now is it fails on linting errors and looks for new errors in the PR. I will keep looking into how to solve these file exists errors

File: utility/types/vote_test.go (hunk 5 above)
Comment (Olshansk) [reply utility/types/vote_test.go]:
```go
	v3 := proto.Clone(v).(*LegacyVote)
	v.BlockHash = []byte("not_a_hash")
```

Screenshot 2023-02-06 at 2 05 51 PM

Comment (h5law) [reply utility/types/vote_test.go]: My bad forgot the pointer 🤣 pushed

File: consensus/module.go [hunk 16]
```diff
@@ -236,15 +238,15 @@ func (m *consensusModule) SetBus(pocketBus modules.Bus) {
...
 	// Sort the validators by their generic param (i.e. service URL)
-	vals := genesis.GetValidators()
+	vals := gen.GetValidators()
 	sort.Slice(vals, func(i, j int) bool {
 		return vals[i].GetGenericParam() < vals[j].GetGenericParam()
 	})
 
 	// Sort the validators by their address
-	vals2 := vals[:]
+	vals2 := vals
```
Comment (Olshansk) [inline consensus/module.go:249]: You will likely need to add a nolint rule here as well.

```suggestion
        // HACK: We're making a copy of `vals` to make sure that the ordering (see comment below) is enforced.
	vals2 := vals[:]
```

**Review by Olshansk — APPROVED** [review]
@h5law I left one comment that requires a minor changes, but otherwise let's get this in so you don't have to cleanup the lints of any other PRs that get merged in first
